File: test_client.py
import argparse
import json
import random
import socket
import threading
import time

NAMES = [
    "Alice", "Bob", "Charlie", "Diana", "Eve", "Frank", "Grace", "Hugo", "Ivy", "Jack",
    "Kara", "Liam", "Mia", "Noah", "Olivia", "Paul", "Quinn", "Ruby", "Sam", "Tina",
    "Uma", "Victor", "Wendy", "Xander", "Yara", "Zane", "Aaron", "Bella", "Caleb", "Daisy",
    "Ethan", "Fiona", "Gavin", "Hannah", "Isaac", "Jade", "Kyle", "Luna", "Mason", "Nora",
    "Owen", "Piper", "Quentin", "Riley", "Sophia", "Tyler", "Ulysses", "Violet", "Wyatt", "Zoe"
]

def random_name():
    return random.choice(NAMES)

def random_gender():
    return random.choice(["male", "female"])

class Client:
    def __init__(self, host, port, name, label, gender=None):
        self.host = host
        self.port = port
        self.name = name
        self.label = label
        self.gender = gender or random_gender()
        self.sock = None
        self.buffer = b""
        self.running = False
        self.thread = None
        self.lock = threading.Lock()
        self.inbox = []
        self.current_qid = None

    def log(self, msg):
        print(f"[{self.label}] {msg}", flush=True)

    def connect(self, send_name=True):
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.sock.connect((self.host, self.port))
        self.running = True
        self.thread = threading.Thread(target=self.recv_loop, daemon=True)
        self.thread.start()
        if send_name:
            self.send({"type": "name", "variable": self.name, "gender": self.gender})
            self.log(f"connected as {self.name} ({self.gender})")
            time.sleep(0.05)
        else:
            self.log("connected without name")

    def close(self):
        self.running = False
        try:
            self.send({"type": "exit"})
        except OSError:
            pass
        try:
            self.sock.close()
        except OSError:
            pass

    def send(self, obj):
        data = json.dumps(obj).encode() + b"\n"
        self.sock.sendall(data)

    def recv_loop(self):
        while self.running:
            try:
                data = self.sock.recv(1024)
            except OSError:
                break
            if not data:
                break
            self.buffer += data
            while b"\n" in self.buffer:
                line, self.buffer = self.buffer.split(b"\n", 1)
                if not line:
                    continue
                try:
                    msg = json.loads(line.decode())
                except json.JSONDecodeError:
                    self.log("bad json from server")
                    continue
                if msg.get("type") == "question":
                    self.current_qid = msg.get("questionId")
                with self.lock:
                    self.inbox.append(msg)
                self.log(f"recv {msg}")

    def wait_for(self, mtype, timeout_s=5):
        deadline = time.time() + timeout_s
        while time.time() < deadline:
            with self.lock:
                for i, msg in enumerate(self.inbox):
                    if msg.get("type") == mtype:
                        return self.inbox.pop(i)
            time.sleep(0.05)
        return None

def assert_true(label, cond, details=""):
    if cond:
        print(f"[PASS] {label}", flush=True)
    else:
        print(f"[FAIL] {label} {details}", flush=True)

def safe_error(client, timeout_s=5):
    err = client.wait_for("error", timeout_s=timeout_s)
    return err.get("message") if err else None

def pregame_tests(host, port, max_clients):
    print("[TEST] no_players")
    ctrl = Client(host, port, "CONTROL", "CTRL")
    ctrl.connect(send_name=False)
    ctrl.send({"type": "start"})
    err_msg = safe_error(ctrl, timeout_s=5)
    assert_true("no_players", err_msg == "no_players")

    print("[TEST] join_ok + players + scoreboard")
    ctrl.send({"type": "name", "variable": ctrl.name, "gender": ctrl.gender})
    assert_true("join_ok", ctrl.wait_for("join_ok") is not None)
    assert_true("players_received", ctrl.wait_for("players") is not None)
    assert_true("scoreboard_received", ctrl.wait_for("scoreboard") is not None)
    time.sleep(0.1)

    print("[TEST] answer_missing / bad_answer_format (no game)")
    ctrl.send({"type": "answer"})
    assert_true("answer_missing", safe_error(ctrl) == "answer_missing")
    time.sleep(0.05)
    ctrl.send({"type": "answer", "choice": "Z"})
    assert_true("bad_answer_format", safe_error(ctrl) == "bad_answer_format")
    time.sleep(0.05)

    print("[TEST] not_joined (no game)")
    ghost = Client(host, port, "GHOST", "GHOST")
    ghost.connect(send_name=False)
    ghost.send({"type": "answer", "choice": "A"})
    ack = ghost.wait_for("answer_ack")
    assert_true("not_joined", ack and ack.get("reason") == "not_joined")
    ghost.close()

    print("[TEST] server_full")
    extras = []
    for i in range(max_clients + 1):
        c = Client(host, port, random_name(), f"E{i}")
        c.connect()
        extras.append(c)
        time.sleep(0.1)
    server_full = False
    for c in extras:
        if safe_error(c, timeout_s=2) == "server_full":
            server_full = True
            break
    assert_true("server_full", server_full)
    for c in extras:
        c.close()

    return ctrl

def manual_session(ctrl, host, port):
    print("[MANUAL] start the game yourself")
    print("Commands: start | A/B/C/D | exit")
    late_bot_done = False
    answered_rounds = 0
    try:
        while True:
            msg = ctrl.wait_for("question", timeout_s=0.1)
            if msg:
                answered_rounds += 1
                if answered_rounds == 3 and not late_bot_done:
                    late = Client(host, port, "LATE_JOIN", "LATE_JOIN")
                    late.connect()
                    time.sleep(0.2)
                    late.send({"type": "answer", "choice": random.choice(["A", "B", "C", "D"])})
                    late.close()
                    late_bot_done = True

            cmd = input("> ").strip().upper()
            if cmd == "START":
                ctrl.send({"type": "start"})
            elif cmd in ["A", "B", "C", "D"]:
                ctrl.send({"type": "answer", "choice": cmd})
            elif cmd == "EXIT":
                break
            elif cmd == "":
                continue
            else:
                print("Unknown command")
    except KeyboardInterrupt:
        pass

def main():
    parser = argparse.ArgumentParser(description="Quiz server pregame test + manual play")
    parser.add_argument("host", nargs="?", default="192.168.4.1", help="Server IP")
    parser.add_argument("--port", type=int, default=7777)
    parser.add_argument("--max-clients", type=int, default=10)
    args = parser.parse_args()

    ctrl = pregame_tests(args.host, args.port, args.max_clients)
    manual_session(ctrl, args.host, args.port)
    ctrl.close()

if __name__ == "__main__":
    main()

File: test_client.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using System.Threading;

public class Client
{
    private static readonly Random Rng = new Random();

    public string Host { get; private set; }
    public int Port { get; private set; }
    public string Name { get; private set; }
    public string Label { get; private set; }
    public string Gender { get; private set; }
    public TcpClient Sock { get; private set; }
    private readonly List<Dictionary<string, string>> _inbox = new List<Dictionary<string, string>>();
    private readonly object _lock = new object();
    private StringBuilder _buffer = new StringBuilder();
    private bool _running = false;
    private Thread _thread;

    public int? CurrentQid { get; private set; }

    public Client(string host, int port, string name, string label, string gender = null)
    {
        Host = host;
        Port = port;
        Name = name;
        Label = label;
        Gender = gender ?? RandomGender();
    }

    public void Log(string msg)
    {
        Console.WriteLine(string.Format("[{0}] {1}", Label, msg));
    }

    public void Connect(bool sendName = true)
    {
        Sock = new TcpClient();
        Sock.Connect(Host, Port);
        _running = true;
        _thread = new Thread(RecvLoop) { IsBackground = true };
        _thread.Start();
        if (sendName)
        {
            Send(new Dictionary<string, object> { ["type"] = "name", ["variable"] = Name, ["gender"] = Gender });
            Log(string.Format("connected as {0} ({1})", Name, Gender));
            Thread.Sleep(50);
        }
        else
        {
            Log("connected without name");
        }
    }

    public void Close()
    {
        _running = false;
        try
        {
            Send(new Dictionary<string, object> { ["type"] = "exit" });
        }
        catch
        {
        }
        try
        {
            Sock.Close();
        }
        catch
        {
        }
    }

    public void Send(Dictionary<string, object> obj)
    {
        string json = JsonSerializer.Serialize(obj);
        byte[] data = Encoding.UTF8.GetBytes(json + "\n");
        NetworkStream stream = Sock.GetStream();
        stream.Write(data, 0, data.Length);
    }

    private void RecvLoop()
    {
        byte[] recvBuffer = new byte[1024];
        while (_running)
        {
            int read;
            try
            {
                read = Sock.GetStream().Read(recvBuffer, 0, recvBuffer.Length);
            }
            catch
            {
                break;
            }
            if (read <= 0)
            {
                break;
            }
            _buffer.Append(Encoding.UTF8.GetString(recvBuffer, 0, read));
            while (true)
            {
                string current = _buffer.ToString();
                int idx = current.IndexOf('\n');
                if (idx < 0)
                {
                    break;
                }
                string line = current.Substring(0, idx);
                _buffer.Remove(0, idx + 1);
                if (line.Length == 0)
                {
                    continue;
                }

                Dictionary<string, string> msg;
                try
                {
                    msg = ParseMessage(line);
                }
                catch
                {
                    Log("bad json from server");
                    continue;
                }

                if (msg.ContainsKey("type") && msg["type"] == "question")
                {
                    if (msg.ContainsKey("questionId") && int.TryParse(msg["questionId"], out int qid))
                    {
                        CurrentQid = qid;
                    }
                }

                lock (_lock)
                {
                    _inbox.Add(msg);
                }
                Log(string.Format("recv {0}", line));
            }
        }
    }

    public Dictionary<string, string> WaitFor(string mtype, double timeoutS = 5.0)
    {
        double deadline = NowSeconds() + timeoutS;
        while (NowSeconds() < deadline)
        {
            lock (_lock)
            {
                for (int i = 0; i < _inbox.Count; i++)
                {
                    if (_inbox[i].ContainsKey("type") && _inbox[i]["type"] == mtype)
                    {
                        Dictionary<string, string> msg = _inbox[i];
                        _inbox.RemoveAt(i);
                        return msg;
                    }
                }
            }
            Thread.Sleep(50);
        }
        return null;
    }

    private static Dictionary<string, string> ParseMessage(string line)
    {
        Dictionary<string, string> dict = new Dictionary<string, string>();
        using (JsonDocument doc = JsonDocument.Parse(line))
        {
            foreach (JsonProperty prop in doc.RootElement.EnumerateObject())
            {
                if (prop.Value.ValueKind == JsonValueKind.String)
                {
                    dict[prop.Name] = prop.Value.GetString() ?? "";
                }
                else
                {
                    dict[prop.Name] = prop.Value.GetRawText();
                }
            }
        }
        return dict;
    }

    private static string RandomGender()
    {
        return Rng.Next(0, 2) == 0 ? "male" : "female";
    }

    private static double NowSeconds()
    {
        return DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0;
    }
}

public static class TestClientProgram
{
    private static readonly List<string> NAMES = new List<string>
    {
        "Alice", "Bob", "Charlie", "Diana", "Eve", "Frank", "Grace", "Hugo", "Ivy", "Jack",
        "Kara", "Liam", "Mia", "Noah", "Olivia", "Paul", "Quinn", "Ruby", "Sam", "Tina",
        "Uma", "Victor", "Wendy", "Xander", "Yara", "Zane", "Aaron", "Bella", "Caleb", "Daisy",
        "Ethan", "Fiona", "Gavin", "Hannah", "Isaac", "Jade", "Kyle", "Luna", "Mason", "Nora",
        "Owen", "Piper", "Quentin", "Riley", "Sophia", "Tyler", "Ulysses", "Violet", "Wyatt", "Zoe"
    };

    private static readonly Random Rng = new Random();

    private static string RandomName()
    {
        return NAMES[Rng.Next(NAMES.Count)];
    }

    private static void AssertTrue(string label, bool cond, string details = "")
    {
        if (cond)
        {
            Console.WriteLine(string.Format("[PASS] {0}", label));
        }
        else
        {
            Console.WriteLine(string.Format("[FAIL] {0} {1}", label, details));
        }
    }

    private static string SafeError(Client client, double timeoutS = 5.0)
    {
        Dictionary<string, string> err = client.WaitFor("error", timeoutS);
        if (err == null)
        {
            return null;
        }
        return err.ContainsKey("message") ? err["message"] : null;
    }

    private static Client PregameTests(string host, int port, int maxClients)
    {
        Console.WriteLine("[TEST] no_players");
        Client ctrl = new Client(host, port, "CONTROL", "CTRL");
        ctrl.Connect(sendName: false);
        ctrl.Send(new Dictionary<string, object> { ["type"] = "start" });
        string errMsg = SafeError(ctrl, 5.0);
        AssertTrue("no_players", errMsg == "no_players");

        Console.WriteLine("[TEST] join_ok + players + scoreboard");
        ctrl.Send(new Dictionary<string, object> { ["type"] = "name", ["variable"] = ctrl.Name, ["gender"] = ctrl.Gender });
        AssertTrue("join_ok", ctrl.WaitFor("join_ok") != null);
        AssertTrue("players_received", ctrl.WaitFor("players") != null);
        AssertTrue("scoreboard_received", ctrl.WaitFor("scoreboard") != null);
        Thread.Sleep(100);

        Console.WriteLine("[TEST] answer_missing / bad_answer_format (no game)");
        ctrl.Send(new Dictionary<string, object> { ["type"] = "answer" });
        AssertTrue("answer_missing", SafeError(ctrl) == "answer_missing");
        Thread.Sleep(50);
        ctrl.Send(new Dictionary<string, object> { ["type"] = "answer", ["choice"] = "Z" });
        AssertTrue("bad_answer_format", SafeError(ctrl) == "bad_answer_format");
        Thread.Sleep(50);

        Console.WriteLine("[TEST] not_joined (no game)");
        Client ghost = new Client(host, port, "GHOST", "GHOST");
        ghost.Connect(sendName: false);
        ghost.Send(new Dictionary<string, object> { ["type"] = "answer", ["choice"] = "A" });
        Dictionary<string, string> ack = ghost.WaitFor("answer_ack");
        AssertTrue("not_joined", ack != null && ack.ContainsKey("reason") && ack["reason"] == "not_joined");
        ghost.Close();

        Console.WriteLine("[TEST] server_full");
        List<Client> extras = new List<Client>();
        for (int i = 0; i < maxClients + 1; i++)
        {
            Client c = new Client(host, port, RandomName(), string.Format("E{0}", i));
            c.Connect();
            extras.Add(c);
            Thread.Sleep(100);
        }
        bool serverFull = false;
        foreach (Client c in extras)
        {
            if (SafeError(c, 2.0) == "server_full")
            {
                serverFull = true;
                break;
            }
        }
        AssertTrue("server_full", serverFull);
        foreach (Client c in extras)
        {
            c.Close();
        }

        return ctrl;
    }
    private static void ManualSession(Client ctrl, string host, int port)
    {
        Console.WriteLine("[MANUAL] start the game yourself");
        Console.WriteLine("Commands: start | A/B/C/D | exit");
        bool lateBotDone = false;
        int answeredRounds = 0;
        try
        {
            while (true)
            {
                Dictionary<string, string> msg = ctrl.WaitFor("question", 0.1);
                if (msg != null)
                {
                    answeredRounds += 1;
                    if (answeredRounds == 3 && !lateBotDone)
                    {
                        Client late = new Client(host, port, "LATE_JOIN", "LATE_JOIN");
                        late.Connect();
                        Thread.Sleep(200);
                        string[] choices = new[] { "A", "B", "C", "D" };
                        late.Send(new Dictionary<string, object> { ["type"] = "answer", ["choice"] = choices[Rng.Next(choices.Length)] });
                        late.Close();
                        lateBotDone = true;
                    }
                }

                string cmd = Console.ReadLine();
                if (cmd == null)
                {
                    continue;
                }
                cmd = cmd.Trim().ToUpperInvariant();
                if (cmd == "START")
                {
                    ctrl.Send(new Dictionary<string, object> { ["type"] = "start" });
                }
                else if (cmd == "A" || cmd == "B" || cmd == "C" || cmd == "D")
                {
                    ctrl.Send(new Dictionary<string, object> { ["type"] = "answer", ["choice"] = cmd });
                }
                else if (cmd == "EXIT")
                {
                    break;
                }
                else if (cmd == "")
                {
                    continue;
                }
                else
                {
                    Console.WriteLine("Unknown command");
                }
            }
        }
        catch (ThreadInterruptedException)
        {
        }
    }

    private static int GetArgInt(string[] args, string name, int fallback)
    {
        for (int i = 0; i < args.Length; i++)
        {
            if (args[i] == name && i + 1 < args.Length && int.TryParse(args[i + 1], out int val))
            {
                return val;
            }
        }
        return fallback;
    }

    public static void Main(string[] args)
    {
        string host = "192.168.4.1";
        if (args.Length > 0 && !args[0].StartsWith("--"))
        {
            host = args[0];
        }
        int port = GetArgInt(args, "--port", 7777);
        int maxClients = GetArgInt(args, "--max-clients", 10);

        Client ctrl = PregameTests(host, port, maxClients);
        ManualSession(ctrl, host, port);
        ctrl.Close();
    }
}
